simpleline/render/widget_containers.py

```diff
@@ -0,0 +1,349 @@
+
+    Widgets will be placed based on the number of columns in the following way:
+
+    w1   w4   w7
+    w2   w5   w8
+    w3   w6   w9
+    """
+
+    def _get_ordered_map(self):
+        ordering_map = self._prepare_list()
+        items_in_column = ceil(len(self._widgets) / self._columns)
+
+        for item_id in range(self.size):
+            col_position = int(item_id // items_in_column)
+            ordering_map[col_position].append(item_id)
+
+        return ordering_map
+
+
+class KeyPattern(object):
+    """Pattern for automatic key printing before items."""
+
+    def __init__(self, pattern="{:d}. ", offset=1, input_description="Press to activate"):
+        """Create the pattern class.
+
+        :param pattern: Set pattern which will be called for every item.
+        :type pattern: Strings format method. See https://docs.python.org/3.3/library/string.html#format-string-syntax.
+
+        :param offset: Set the offset for numbering items. Default is 1 to start indexing naturally for user.
+        :type offset: int
+
+        :param input_description: Input description for user. What the user is activating. This will be printed in
+                                  prompt.
+        :type input_description: str
+        """
+        self._pattern = pattern
+        self._offset = offset
+        self._input_description = input_description
+
+    def get_prompt_help_and_description(self, max_item_id):
+        """Get user help to activate items in the list.
+
+        What the user is doing by activating these items.
+
+        :param max_item_id: ID of the last item.
+        :type max_item_id: int
+
+        :returns: Return (key hint, description).
+                  * key hint is what user must type to activate items
+                  * description what the user is doing by typing 'key hint'
+        :rtype: (str, str)
+        """
+        user_input_help = "1 - %s" % (max_item_id + self._offset)
+        description = self._input_description
+        result = (user_input_help, description)
+        return result
+
+    def get_widget_key_from_id(self, item_id):
+        """Get widget identifier for user input description.
+
+        It should be something similar to the pattern.
+
+        :param item_id: Position of the widget in the list.
+        :type item_id: int starts from 0.
+        """
+        return self._pattern.format(item_id + self._offset)
+
+    def get_widget_id_from_key(self, user_input):
+        """Get id of the widget from user input.
+
+        This is reverse translation to `self.get_widget_identifier()`.
+
+        :param user_input: Input from user:
+        :type user_input: str
+
+        :return: ID of the widget in the list or None if the input can't be translated.
+        :rtype: int or None
+        """
+        try:
+            return int(user_input) - 1
+        except ValueError:
+            print("Value Error")
+            return None
+
+
+class WidgetContainer(object):
```

Comment:
When I talked about containers, I meant `ListRowWidget` and `ListColumntWidget`, because they contain other widgets (see https://developer.gnome.org/gtk3/stable/GtkContainer.html). It is confusing to use the name `WidgetContainer` for something what is more like `ListItem`. 

Also, I don't like `WidgetContainer` at all. It should be used only privately by `List*Widgets`, but we shouldn't be forced to use it since it doesn't bring any extra value (it is just a holder for data). I saw the tests and just the fact, that you call `make_container_list`, suggests that this is a bad decision.

It looks like what we are trying to do here is to introduce a modifiable widget. However, the original goal was to simplify `input` and `refresh` methods in spokes and this change doesn't serve the purpose. If we really need modifiable widgets, lets make the `Widget` classes modifiable, but not this way. Nevertheless, we haven't need them to be modifiable so far. 

I know we talked about this already, so I am sorry that I wasn't clearer before.

What I would like to do in spokes:
```      
class SomeSpoke(NormalTUISpoke):
  
  def refresh(self, args=None):
    # Create list of options.
    container = ListRowContainer(...)
    container.add(TextWidget("A"), self._process_a, data_a)
    container.add(TextWidget("B"), self._process_b, data_b)

    if flags.option_c_is_allowed:
      container.add(TextWidget("C"), self._process_c, data_c)

    # Create window.
    self.window = WindowContainer(self.title)
    self.window.add(container)
    self.window.add_space()

  def input(self, args, key):
    if self.container.process_input(args, key):
      return INPUT_PROCESSED

    if key == "Q":
      # Process somehow this extra option

    return INPUT_DISCARDED

  def _process_a(self, data_a):
    # Option A was selected, process the data.
    pass

  def _process_b(self, data_b):
    # Option B was selected, process the data.
    pass

  def _process_c(self, data_c):
    # Option C was selected, process the data.
    pass

```